generator/src/AbstractGenerator.php

```diff
@@ -0,0 +1,127 @@
+use function sort;
+use function ucfirst;
+
+/** @internal */
+abstract class AbstractGenerator
+{
+    /** @var array<string, list<string|class-string>> */
+    protected array $typeAliases = [
+        'resolvesToExpression' => [Expression\ResolvesToExpression::class, 'array', 'object', 'string', 'int', 'float', 'bool', 'null'],
+        'resolvesToArrayExpression' => [Expression\ResolvesToArrayExpression::class, 'array', 'object', 'string'],
+        'resolvesToBoolExpression' => [Expression\ResolvesToBoolExpression::class, 'array', 'object', 'string', 'bool'],
+        'resolvesToMatchExpression' => ['array', 'object', Expression\ResolvesToMatchExpression::class],
+        'resolvesToNumberExpression' => [Expression\ResolvesToBoolExpression::class, 'array', 'object', 'string', 'int', 'float'],
+        'resolvesToQueryOperator' => ['array', 'object', Expression\ResolvesToQueryOperator::class],
+        'resolvesToSortSpecification' => ['array', 'object', Expression\ResolvesToSortSpecification::class],
+    ];
+
+    protected GeneratorDefinition $definition;
+    protected Printer $printer;
+
+    public function __construct(GeneratorDefinition $definition)
+    {
+        $this->validate($definition);
+
+        $this->definition = $definition;
+        $this->printer = new PsrPrinter();
+    }
+
+    /** @throws InvalidArgumentException when definition is invalid */
+    protected function validate(GeneratorDefinition $definition): void
+    {
+    }
+
+    public function createClassesForObjects(array $objects): void
+    {
+        foreach ($objects as $object) {
+            $this->createFileForClass(
+                $this->definition->filePath,
+                $this->createClassForObject($object),
+            );
+        }
+    }
+
+    abstract public function createClassForObject(object $object): ClassType;
+
+    /** @return array{native:string,doc:string} */
+    final protected function generateTypeString(ArgumentDefinition $arg): array
+    {
+        $type = $arg->type;
+        $nativeTypes = $this->typeAliases[$type] ?? [$type];
+        $docTypes = $nativeTypes;
+
+        foreach ($nativeTypes as $key => $typeName) {
+            if (interface_exists($typeName)) {
+                $nativeTypes[$key] = $docTypes[$key] = '\\' . $typeName;
+                // A union cannot contain both object and a class type, which is redundant and causes a PHP error
+                // @todo replace "object" with "stdClass" and force any class object to implement the proper interface
+                if (in_array('object', $nativeTypes, true)) {
+                    unset($nativeTypes[$key]);
+                }
+            }
+        }
+
+        sort($nativeTypes);
+        sort($docTypes);
+
+        if ($arg->isOptional) {
+            $nativeTypes[] = 'null';
+            $docTypes[] = 'null';
+        }
+
+        return [
+            'native' => implode('|', $nativeTypes),
+            'doc' => implode('|', $docTypes),
```

Comment:
Note: you may want to use `array_unique` here in case we ever start out with a union in the config.

Reply:
Added.